### PR TITLE
DWR-503 Attempt at the loop by partition logic

### PR DESCRIPTION
**Pull request by agorina**
Just want to know what you think about this approach for partitioning and looping.

### CODE REVIEW

File: legacy_load/sql/warehouse_legacy_load.sql [hunk 1]
```diff
@@ -1,26 +1,65 @@
 USE legacy_load;
 
+
+DROP PROCEDURE IF EXISTS loop_by_partition;
```
Comment (agorina) [inline legacy_load/sql/warehouse_legacy_load.sql]: @malaffoon, I know you did not want to use stored procedures, but just in case....check it out.

File: legacy_load/sql/warehouse_legacy_load.sql [hunk 2]
```diff
@@ -1,26 +1,65 @@
...
+
+DELIMITER //
+CREATE PROCEDURE loop_by_partition(IN p_sql VARCHAR(1000), IN p_size INTEGER)
+  BEGIN
+    DECLARE iteration INTEGER;
+    SET iteration = 0;
+
+    label1: LOOP
+      SET iteration = iteration + 1;
+      SET @stmt = concat( p_sql, ' and warehouse_partition_id =', iteration);
```
Comment (agorina) [inline legacy_load/sql/warehouse_legacy_load.sql]: It assumes that a p_sql given in is for a table that has warehouse_partition_id column.

File: legacy_load/sql/warehouse_legacy_load.sql [hunk 3]
```diff
@@ -1,26 +1,65 @@
...
-  JOIN dim_asmt_guid_to_natural_id_mapping m ON m.guid = da.asmt_guid -- TODO: this needs to be adjusted
+  JOIN (select concat('(SBAC)', natural_id, '-Winter-', school_year) as natural_id, guid from  dim_asmt_guid_to_natural_id_mapping ) as m ON m.guid = da.asmt_guid -- TODO: this needs to be adjusted
+  JOIN warehouse.asmt wa ON wa.natural_id = m.natural_id
+SET warehouse_asmt_id = wa.id
+WHERE warehouse_load_id = 33;
+
+# some asmt_guid are not guids
+UPDATE dim_asmt da
+  JOIN (select asmt_rec_id, concat('(SBAC)', asmt_guid, '-Winter-', asmt_period_year, '-', asmt_period_year+1) as natural_id from  dim_asmt ) as m ON m.asmt_rec_id = da.asmt_rec_id -- TODO: this needs to be adjusted
   JOIN warehouse.asmt wa ON wa.natural_id = m.natural_id
```
Comment (agorina) [inline legacy_load/sql/warehouse_legacy_load.sql]: These changes are unrelated to the looping. It is a fix to find matching asmts. 
As I mentioned offline it is still not enough. Will chat tomorrow.

File: legacy_load/sql/warehouse_legacy_load.sql [hunk 4]
```diff
@@ -1,26 +1,65 @@
...
+  SET warehouse_partition_id = MOD(ds.student_rec_id, 5);
+
+CALL loop_by_partition('UPDATE dim_student ds
 SET warehouse_gender_id =
-CASE WHEN ds.sex = 'male' THEN (SELECT id FROM warehouse.gender WHERE code = 'Male')
-ELSE (SELECT id FROM warehouse.gender WHERE CODE = 'Female') END
-WHERE warehouse_load_id = 33;
+CASE WHEN ds.sex = ''male'' THEN (SELECT id FROM warehouse.gender WHERE code = ''Male'')
+ELSE (SELECT id FROM warehouse.gender WHERE CODE = ''Female'') END
+WHERE warehouse_load_id = 33', 5);
```
Comment (agorina) [inline legacy_load/sql/warehouse_legacy_load.sql]: This is an example of the `loop_by_partition` call.

File: legacy_load/sql/warehouse_legacy_load.sql [hunk 5]
```diff
@@ -59,14 +98,11 @@ SELECT exists(SELECT 1
...
 
+
 SELECT exists(SELECT 1
               FROM dim_asmt
               WHERE warehouse_asmt_id IS NULL AND warehouse_load_id = 33);
 
-SELECT exists(SELECT 1
-              FROM dim_student
-              WHERE warehouse_gender_id IS NULL AND warehouse_load_id = 33);
-
```
Comment (agorina) [inline legacy_load/sql/warehouse_legacy_load.sql]: This is useless check.

File: legacy_load/sql/warehouse_legacy_load.sql [hunk 6]
```diff
@@ -232,63 +268,63 @@ INSERT INTO load_progress (warehouse_load_id, message) VALUE (33, 'reset exam_st
 
 INSERT INTO warehouse.exam_available_accommodation (exam_id, accommodation_id)
   SELECT warehouse_exam_id, (SELECT id from warehouse.accommodation where code = 'TDS_ASL1') as  accommodation_id
-  from fact_block_asmt_outcome f where acc_asl_video_embed != 0;
```
Comment (agorina) [inline legacy_load/sql/warehouse_legacy_load.sql]: Unrelated, fix to the accommodations.

File: legacy_load/sql/warehouse_legacy_load.sql [hunk 7]
```diff
@@ -1,26 +1,65 @@
 USE legacy_load;
 
+
+DROP PROCEDURE IF EXISTS loop_by_partition;
+
+DELIMITER //
+CREATE PROCEDURE loop_by_partition(IN p_sql VARCHAR(1000), IN p_size INTEGER)
```
Comment (malaffoon) [inline legacy_load/sql/warehouse_legacy_load.sql]: I'd suggest passing in `IN p_first INTEGER, IN p_last INTEGER` instead. That way, we can easily execute for an offset set of partition ids.
